addon/utils/guess-locale.js

```diff
@@ -0,0 +1,91 @@
+import { A as array } from '@ember/array';
+
+/**
+ * Guess a locale based on allowed & desired locales.
+ * This will return the best-fitting locale.
+ *
+ * Given the following input:
+ * allowedLocales = ['en', 'de', 'zh_HK']
+ * desiredLocales = ['de-AT', 'de', 'en-US', 'en']
+ *
+ * It would by default return 'de'.
+ *
+ * If you specify `allowSubLocales=true`, it would instead return `de_AT`, the favorite sub-locale.
+ *
+ * In contrast, the following input:
+ * allowedLocales = ['en', 'de', 'zh_HK']
+ * desiredLocales = ['zh-CN', 'zh-HK', 'en-US', 'en']
+ *
+ * Would always return 'zh_HK', no matter if sub locales are allowed or not.
+ *
+ * @method guessLocale
+ * @param allowedLocales
+ * @param desiredLocales
+ * @param defaultLocale
+ * @param allowSubLocales
+ * @return {String}
+ */
+export function guessLocale(allowedLocales = [], desiredLocales = [], { defaultLocale = 'en', allowSubLocales = false } = {}) {
+  desiredLocales = desiredLocales || [defaultLocale];
+  desiredLocales = desiredLocales.map(normalizeLocale).map(getLocalAlias);
+
+  // Ensure everything is an Ember Array
+  if (!desiredLocales.find) {
+    desiredLocales = array(desiredLocales);
+  }
+  if (!allowedLocales.find) {
+    allowedLocales = array(allowedLocales);
+  }
+
+  let locale = desiredLocales.find((locale) => {
```

Comment:
But in this case, how could this be re-written to a hash map? Since we are comparing all of these items through a function, trying to find the first fit?

Also, in this case this would certainly be premature optimization, as we're never going to have more than ~100 array items in there, and there will certainly never be a noticeable or maybe even measureable difference.

Reply:
Yeah in this case it's an optimization which probably makes no difference for users. I just wanted to foster awareness that loops in loops can be source of performance problems when collections are large. The same code with a hash map would be:

```
// make map = O(n)
let map = {};
allowedLocales.forEach((l) => { map[l]=1; map[l.substring(0,2)]=1; });

// do the lookup = O(n)
let locale = desiredLocales.find((l) => !!map[l] || !!map[l.substring(0,2)]);
```

That would lead to the same results, but without `O(n^2)` :) But as you said, it's not a problem for small arrays!